plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java

```diff
@@ -438,31 +459,63 @@ private BigQueryOutputTableHandle createTable(ConnectorSession session, Connecto
+                columnsTypes.build(),
+                temporaryTableName,
+                pageSinkIdColumn.map(ColumnMetadata::getName));
+    }
+
+    private TableId createTable(BigQueryClient client, String projectId, String remoteSchemaName, String tableName, ImmutableList<Field> fields, Optional<String> tableComment)
+    {
         TableId tableId = TableId.of(projectId, remoteSchemaName, tableName);
-        TableDefinition tableDefinition = StandardTableDefinition.of(Schema.of(fields.build()));
+        TableDefinition tableDefinition = StandardTableDefinition.of(Schema.of(fields));
         TableInfo.Builder tableInfo = TableInfo.newBuilder(tableId, tableDefinition);
-        tableMetadata.getComment().ifPresent(tableInfo::setDescription);
+        tableComment.ifPresent(tableInfo::setDescription);
 
         client.createTable(tableInfo.build());
 
-        return new BigQueryOutputTableHandle(new RemoteTableName(tableId), columnsNames.build(), columnsTypes.build());
+        return tableId;
     }
 
     @Override
     public Optional<ConnectorOutputMetadata> finishCreateTable(ConnectorSession session, ConnectorOutputTableHandle tableHandle, Collection<Slice> fragments, Collection<ComputedStatistics> computedStatistics)
     {
-        return Optional.empty();
+        BigQueryOutputTableHandle handle = (BigQueryOutputTableHandle) tableHandle;
+        checkState(handle.getTemporaryTableName().isPresent(), "Unexpected use of finishCreateTable without a temporaryTableName present");
+        return finishInsert(session, handle.getRemoteTableName(), handle.getTemporaryRemoteTableName().get(), handle.getPageSinkIdColumnName().get(), handle.getColumnNames(), fragments);
     }
 
     @Override
     public void dropTable(ConnectorSession session, ConnectorTableHandle tableHandle)
     {
-        BigQueryClient client = bigQueryClientFactory.create(session);
         BigQueryTableHandle bigQueryTable = (BigQueryTableHandle) tableHandle;
-        if (isWildcardTable(TableDefinition.Type.valueOf(bigQueryTable.asPlainTable().getType()), bigQueryTable.asPlainTable().getRemoteTableName().getTableName())) {
+        dropTable(session, bigQueryTable.asPlainTable());
+    }
+
+    private void dropTable(ConnectorSession session, BigQueryNamedRelationHandle handle)
+    {
+        if (isWildcardTable(TableDefinition.Type.valueOf(handle.getType()), handle.getRemoteTableName().getTableName())) {
             throw new TrinoException(BIGQUERY_UNSUPPORTED_OPERATION, "This connector does not support dropping wildcard tables");
         }
-        TableId tableId = bigQueryTable.asPlainTable().getRemoteTableName().toTableId();
+        dropTable(session, handle.getRemoteTableName().toTableId());
+    }
+
+    // This is a last resort method - use one of the ones above if you have that type.
+    // Only use this if you are certain the TableId does not refer to a wildcard table.
+    private void dropTable(ConnectorSession session, TableId tableId)
```

Comment:
Calling the factory method in closers is simpler than separating `dropTable` method in my opinion:
```java
closer.register(() -> bigQueryClientFactory.create(session).dropTable(tempTable.toTableId()));
...
closer.register(() -> bigQueryClientFactory.create(session).dropTable(pageSinkTable.toTableId()));
```